adrs/2023-02-10-limit-manager-role-permission.md

```diff
@@ -0,0 +1,79 @@
+
+- The cache-based API client requires cluster scope `list` and `watch` permission for any resource the controller may query.
+
+- This documentation only scopes to the AutoscalingRunnerSet CRD and its controller.
+
+## Service accounts and their role binding in actions-runner-controller
+
+There are 3 service accounts involved for a working `AutoscalingRunnerSet` based `actions-runner-controller`
+
+1. Service account for each Ephemeral runner Pod
+
+This should have the lowest privilege (not any `RoleBinding` nor `ClusterRoleBinding`) by default, in the case of `containerMode=kubernetes`, it will get certain write permission with `RoleBinding` to limit the permission to a single namespace.
+
+> References:
+> - ./charts/auto-scaling-runner-set/templates/no_permission_serviceaccount.yaml
+> - ./charts/auto-scaling-runner-set/templates/kube_mode_role.yaml
+> - ./charts/auto-scaling-runner-set/templates/kube_mode_role_binding.yaml
+> - ./charts/auto-scaling-runner-set/templates/kube_mode_serviceaccount.yaml
+
+2. Service account for AutoScalingListener Pod
+
+This has a `RoleBinding` to a single namespace with a `Role` that has permission to `PATCH` `EphemeralRunnerSet` and `EphemeralRunner`.
+
+3. Service account for the controller manager
+
+Since the CRD controller is a singleton installed in the cluster that manages the CRD across multiple namespaces, the service account of the controller manager pod has a `ClusterRoleBinding` to a `ClusterRole` with broader permissions.
+
+The current `ClusterRole` has the following permissions:
+
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingListeners` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunners` (with `Status` and `Finalizer` sub-resource)
+
+- Get/List/Create/Delete/Update/Patch/Watch on `Pods` (with `Status` sub-resource)
+- **Get/List/Create/Delete/Update/Patch/Watch on `Secrets`**
+- Get/List/Create/Delete/Update/Patch/Watch on `Roles`
+- Get/List/Create/Delete/Update/Patch/Watch on `RoleBindings`
+- Get/List/Create/Delete/Update/Patch/Watch on `ServiceAccounts`
+
+> Full list can be found at: https://github.com/actions/actions-runner-controller/blob/facae69e0b189d3b5dd659f36df8a829516d2896/charts/actions-runner-controller-2/templates/manager_role.yaml
+
+## Limit cluster role permission on Secrets
+
+The cluster scope `List` `Secrets` permission might be a blocker for adopting `actions-runner-controller` for certain customers as they may have certain restriction in their cluster that simply doesn't allow any service account to have cluster scope `List Secrets` permission. 
```

Comment:
I'm not representing all the customers who would love to use ARC, but they are constrained by tighter security rules, but I can add my 5 cents.

This section starts with "The cluster scope `List` `Secrets` permission.. ".  I don't think `List` or in fact `Watch` on the cluster level is a problem here. I'm a lot more concerned about all the rest of them: `Get/Create/Delete/Update/Patch`. There are the ones which can create havoc on the entire cluster.

Reply:
@wherka-ama do you concern about the permission that ARC itself has or the permission each runner pod has? The ARC is open-sourced, so you can audit what operation it's doing with its permission.

Reply:
> @wherka-ama do you concern about the permission that ARC itself has or the permission each runner pod has? The ARC is open-sourced, so you can audit what operation it's doing with its permission.

@TingluoHuang my concern is not really about ARC itself. I'm perfectly happy with the fact it's open. Let's not change that. I love open source and the fact I can easily refer to source code when the docs or the behaviour of the software don't make perfect sense right away. However, the provenance aka supply chain needs to be assured to build that trust. From GO sources, via binaries, to docker image there is a long way and we need to be sure that this chain is never tampered with. I hope we will not need to worry about these things here. Just stating the obvious.

My concern is not about ARC's internal logic - more about the doors which are opened for it and not secured which may lead to use them for malicious purposes.  It is that indirect access that gives us headache and principle of least privilege, which is not necessarily followed religiously in ARC. 

You can imagine the case where the admin user installs ARC with all its ClusterRoles, ClusterRoleBindings and binds them to a specific service account which is typically used to run the business of ARC manager. 
As soon as one can use such service account - completely independently of ARC - it's as easy as starting the POD with with curl or a kubectl inside, pointing to certificate automatically mounted for a service account in question and completely impersonating it. You can only imagine what kind of operation you could perform on it's behalf. Starting from reading all the secrets on the cluster, then modifying them or deleting, because you can, going further to granting yourself new clusterroles and binding them on the cluster level.. that's pretty much unrestricted access to the entire cluster. 
It's not very hard to imagine someone using a simple recipe like this:
https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/#accessing-the-api-from-within-a-pod